include/revng/ADT/GenericGraph.h

```diff
@@ -388,6 +390,488 @@ class BidirectionalNode
+
+  using PredecessorEdgeIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToView>;
+  using ConstPredecessorEdgeIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToConstView>;
+  using PredecessorIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToNeighbor>;
+  using ConstPredecessorIterator = revng::mapped_iterator<
+    NonOwningEdge *,
+    typename PredecessorFilters::Types::ToConstNeighbor>;
+
+public:
+  llvm::iterator_range<SuccessorEdgeIterator> successor_edges() {
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return revng::map_range(Range, SuccessorFilters::ToView);
+  }
+  llvm::iterator_range<ConstSuccessorEdgeIterator> successor_edges() const {
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return revng::map_range(Range, SuccessorFilters::ToConstView);
+  }
+  llvm::iterator_range<SuccessorIterator> successors() {
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return revng::map_range(Range, SuccessorFilters::ToNeighbor);
+  }
+  llvm::iterator_range<ConstSuccessorIterator> successors() const {
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return revng::map_range(Range, SuccessorFilters::ToConstNeighbor);
+  }
+
+  llvm::iterator_range<PredecessorEdgeIterator> predecessor_edges() {
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return revng::map_range(Range, PredecessorFilters::ToView);
+  }
+  llvm::iterator_range<ConstPredecessorEdgeIterator> predecessor_edges() const {
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return revng::map_range(Range, PredecessorFilters::ToConstView);
+  }
+  llvm::iterator_range<PredecessorIterator> predecessors() {
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return revng::map_range(Range, SuccessorFilters::ToNeighbor);
+  }
+  llvm::iterator_range<ConstPredecessorIterator> predecessors() const {
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return revng::map_range(Range, SuccessorFilters::ToConstNeighbor);
+  }
+
+protected:
+  auto findSuccessorImpl(DerivedType const &S) {
+    auto Comparator = [&S](auto &Edge) { return Edge.Neighbor == &S; };
+    return std::find_if(Successors.begin(), Successors.end(), Comparator);
+  }
+  auto findSuccessorImpl(DerivedType const &S) const {
+    auto Comparator = [&S](auto const &Edge) { return Edge.Neighbor == &S; };
+    return std::find_if(Successors.begin(), Successors.end(), Comparator);
+  }
+  auto findPredecessorImpl(DerivedType const &P) {
+    auto Comparator = [&P](auto &Edge) { return Edge.Neighbor == &P; };
+    return std::find_if(Predecessors.begin(), Predecessors.end(), Comparator);
```

Comment:
So you're preventing from having more than one edge between two nodes, even if they have different labels?
Fair, if it fits your use case. We can relax this in the future if we need to.
Make sure to document this clearly.

Reply:
I think we have to do it this way, because our labels are mutable, nothing prevents user from change a label later to cause duplication. And controlling every single label change (moreover with a linear search in case it caused a duplication) doesn't sound like a good idea.

My vote's for behavior like this: a pair of nodes either doesn't have an edge, or has exactly one.

Reply:
Seems appropriate.